vnext/Microsoft.ReactNative/JsiReader.cpp

```diff
@@ -0,0 +1,133 @@
+
+      // unfortunately JSI doesn't differentiate int and double
+      // here we test if the double value can be converted to int without data loss
+      // treat it like an int if we succeeded
+
+      if (floor(number) == number) {
+        return JSValueType::Int64;
+      } else {
+        return JSValueType::Double;
+      }
+    }
+  } else if (m_containers.size() > 0) {
+    return m_containers[m_containers.size() - 1].Type == ContainerType::Object ? JSValueType::Object
+                                                                               : JSValueType::Array;
+  }
+  return JSValueType::Null;
+}
+
+bool JsiReader::GetNextObjectProperty(hstring &propertyName) noexcept {
+  if (m_containers.size() == 0) {
+    return false;
+  }
+
+  auto &top = m_containers[m_containers.size() - 1];
+  if (top.Type != ContainerType::Object) {
+    return false;
+  }
+
+  top.Index++;
+  if (top.Index < static_cast<int>(top.PropertyNames.value().size(m_runtime))) {
+    auto propertyId =
+        top.PropertyNames.value().getValueAtIndex(m_runtime, static_cast<size_t>(top.Index)).getString(m_runtime);
+    propertyName = winrt::to_hstring(propertyId.utf8(m_runtime));
+    SetValue(top.CurrentObject.value().getProperty(m_runtime, propertyId));
+    return true;
+  } else {
+    m_containers.pop_back();
+    return false;
+  }
+}
+
+bool JsiReader::GetNextArrayItem() noexcept {
+  if (m_containers.size() == 0) {
+    return false;
+  }
+
+  auto &top = m_containers[m_containers.size() - 1];
+  if (top.Type != ContainerType::Array) {
+    return false;
+  }
+
+  top.Index++;
+  if (top.Index < static_cast<int>(top.CurrentArray.value().size(m_runtime))) {
+    SetValue(top.CurrentArray.value().getValueAtIndex(m_runtime, static_cast<size_t>(top.Index)));
+    return true;
+  } else {
+    m_containers.pop_back();
+    return false;
+  }
+}
+
+hstring JsiReader::GetString() noexcept {
+  if (ValueType() != JSValueType::String) {
+    return {};
+  }
+  return winrt::to_hstring(m_currentPrimitiveValue.value().getString(m_runtime).utf8(m_runtime));
+}
+
+bool JsiReader::GetBoolean() noexcept {
+  if (ValueType() != JSValueType::Boolean) {
+    return false;
+  }
+  return m_currentPrimitiveValue.value().getBool();
+}
+
+int64_t JsiReader::GetInt64() noexcept {
+  if (ValueType() != JSValueType::Int64) {
+    return 0;
+  }
+  return static_cast<int64_t>(m_currentPrimitiveValue.value().getNumber());
+}
+
+double JsiReader::GetDouble() noexcept {
+  auto valueType = ValueType();
+  if (valueType != JSValueType::Int64 && valueType != JSValueType::Double) {
+    return 0;
+  }
+  return m_currentPrimitiveValue.value().getNumber();
+}
+
+void JsiReader::SetValue(const facebook::jsi::Value &value) noexcept {
+  if (value.isObject()) {
+    auto obj = value.getObject(m_runtime);
+    if (obj.isArray(m_runtime)) {
+      m_containers.push_back(obj.getArray(m_runtime));
+    } else {
+      m_containers.push_back({m_runtime, std::move(obj)});
```

Comment:
not sure about the design, but functions would fall into this case as well. Should we have a check for that?
i.e. use:

```
  /// \return true iff the Object is callable.  If so, then \c
  /// getFunction will succeed.
  bool isFunction(Runtime& runtime) const {
    return runtime.isFunction(*this);
  }
```

Reply:
In this case properties of Function will get enumerated. There are still other cases, you are not able to check all system classes. But this is fine, if users use our codegen, they won't be able to do this in TypeScript.